### PR TITLE
TS-4265: Restructure thread initialization

### PR DESCRIPTION
**Pull request by SolidWallOfCode**
This is primarily a precursor to other changes which are too large for a single PR. It does have some (IMHO) significant benefits of itself.

* Thread initialization is no longer subject to race conditions or "getting lucky" on the timing, where the current code starts the threads *and then initializes them*. Seriously. This changes that to allow thread startup continuations to be registered and executed *in the target thread* while the thread is starting up.
* This also enables decoupling and a reduction of circular dependencies. The thread start up logic no longer needs to be aware of how a client module needs to have a thread initialized. The client module registers a start up continuation which the event system simply executes at thread start just like any other continuation.
* The event system logic is now at least partially documented both in code comments and the official documentation.
* Information about threads and thread groups is in a structure instead of being parallel arrays.
* A fix for thread start up race conditions is present which IMHO is better than #1989.

P.S. @PSUdaemon pre-reviewed this. I integrated all of his requested changes except one because that change causes a compilation failure.

### CODE REVIEW

File: iocore/eventsystem/I_EventProcessor.h [hunk 1]
```diff
@@ -30,15 +30,15 @@
...
-const int MAX_THREADS_IN_EACH_TYPE = 3072;
+constexpr int MAX_THREADS_IN_EACH_TYPE = 3072;
 #endif
 
 #ifdef TS_MAX_NUMBER_EVENT_THREADS
-const int MAX_EVENT_THREADS = TS_MAX_NUMBER_EVENT_THREADS;
+constexpr int MAX_EVENT_THREADS = TS_MAX_NUMBER_EVENT_THREADS;
 #else
-const int MAX_EVENT_THREADS = 4096;
+constxpr int MAX_EVENT_THREADS = 4096;
```
Comment (bryancall) [inline iocore/eventsystem/I_EventProcessor.h]: Spelled wrong

Comment (SolidWallOfCode) [inline iocore/eventsystem/I_EventProcessor.h]: I'll drop that in when I get a fix for the FreeBSD crash. But first I need FreeBSD somewhere.

**Comment by SolidWallOfCode** [comment]
@jpeach Haha, this is *already* a broken up commit! It's just phase 1 of 4 or 5  I had planned (including continuation tracking and reworking the event loop logic). See TS_4260 for an earlier plan for this overall project.

I could break out the stack size as a compile time option and the member initialization, but changing the signature of `spawn_event_threads` is intimately tied to the other changes, as it changing the signature of `spawn_event_threads`.

I can't really move the documentation to another commit because it documents the new structure, not the old one. I am definitely not going to write documentation for the old style so I can replace it with updated documentation when this gets committed.

I'll see what I can do.

**Comment by SolidWallOfCode** [comment]
[approve ci]

File: doc/developer-guide/architecture/threads-and-events.en.rst [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+:class:`EThread` is a subclass of :class:`Thread` which provides support for |TS| core operations. It is this class that provides support for using :class:`Continuation` instances. Each :class:`EThread` instance executes a single continuation at thread start. If the thread is :enumerator:`ThreadType::DEDICATED` it exits after invoking the start continuation. A :enumerator:`ThreadType::REGULAR` thread will process its event queue until explicitly stopped after executing the start continuation.
```
Comment (oknet) [inline doc/developer-guide/architecture/threads-and-events.en.rst]: >> If the thread is :enumerator:`ThreadType::DEDICATED` it exits after invoking the start continuation.

The DEDICATED EThread is not exits since no one calls pthread_join().
It is just stops and keeps its stack like a zombie process until the ATS process exits.

**Comment by SolidWallOfCode** [comment]
I broke out some of the changes, waiting on those PRs to get approved.

**Comment by SolidWallOfCode** [comment]
Updated with all recommendations implemented.

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

File: doc/developer-guide/architecture/threads-and-events.en.rst [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+:class:`EventProcessor` is intended to be a singleton and the global instance is :var:`eventProcessor`.
+
+In general if a subsystem in the |TS| core is setting up a thread group, it should use code of the form ::
+
+   int ET_GROUP; // global variable, where "GROUP" is repalced by the actual group / type name.
+   int n_group_threads = 3; // Want 3 of these threads by default, possibly changed by configuration options.
+   constexpr size_t GROUP_STACK_SIZE = 1 << 20; // stack size for each thread.
+   void Group_Thread_Init(EThread*); // function to perform per thread local initialization.
+   //
```
Comment (jpeach) [inline doc/developer-guide/architecture/threads-and-events.en.rst]: Leave this line blank.

File: doc/developer-guide/architecture/threads-and-events.en.rst [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+   ET_GROUP = eventProcessor::registerEventType("Group");
+   eventProcessor.schedule_spawn(&Group_Per_Thread_Init, ET_GROUP);
+   eventProcessor.spawn_event_threads(ET_GROUP, n_group_threads, GROUP_STACK_SIZE);
```
Comment (jpeach) [inline doc/developer-guide/architecture/threads-and-events.en.rst]: The code example should be in a code block:
```
.. code-block:: cpp
```

File: doc/developer-guide/architecture/threads-and-events.en.rst [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+
+The function :code:`Group_Thread_Init` can be replaced with a continuation if that's more convenient. One advantage of a continuation is additional data (via :arg:`cookie`) can be provide during thread initialization.
+
+If there is no thread initializatoin needed, this can be compressed in to a single call ::
+
+   ET_GROUP = eventProcessor.spawn_event_threads("Group", n_group_threads, GROUP_STACK_SIZE);
```
Comment (jpeach) [inline doc/developer-guide/architecture/threads-and-events.en.rst]: Put this in a case block.

File: doc/developer-guide/architecture/threads-and-events.en.rst [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+This registers the group name and type, starts the threads, and returns the event type.
+
+TS-4265 Changes
```
Comment (jpeach) [inline doc/developer-guide/architecture/threads-and-events.en.rst]: I don't think we should have permanent documentation of bug numbers from a deprecated system. Do we need this paragraph? We should just describe the system as it works now.

File: doc/developer-guide/architecture/threads-and-events.en.rst [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+Types
+=====
+
+.. type:: EventType
+
+   A thread classification value that represents the type of events the thread is expected to process.
+
+.. var:: EventType ET_CALL
+
+   A predefined :type:`EventType` which always exists. This is deprecated, use :var:`ET_NET` instead.
```
Comment (jpeach) [inline doc/developer-guide/architecture/threads-and-events.en.rst]: It is deprecated? That ought to be in a code comment or compiler annotation if that's so.

File: doc/developer-guide/architecture/threads-and-events.en.rst [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+   referenced by an :class:`Action` instance. This is used primarily to allow the future work to be
+   canceled.
+
+.. class:: Action
+
+   Reference to a future computation for a :class:`Continuation`.
+
+.. class:: Event : public Action
+
+   Reference to code to dispatch. Note that an :class:`Event` is a type of :class:`Action`. This class combines the future computational reference of :class:`Action`
```
Comment (jpeach) [inline doc/developer-guide/architecture/threads-and-events.en.rst]: Line wrap.

File: doc/developer-guide/architecture/threads-and-events.en.rst [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+      A thread which executes only the start contiuation and then exits.
+
+   .. enumerator:: REGULAR
+
+      A thread which executes the start continuation and then processes its event queue.
+
+   .. enumerator:: MONITOR
+
+      Unknown - does not appear to be used, should be removed entirely.
```
Comment (jpeach) [inline doc/developer-guide/architecture/threads-and-events.en.rst]: You removed this in this patch, why document it?

File: iocore/eventsystem/I_Thread.h [hunk 10]
```diff
@@ -75,8 +75,8 @@ constexpr int MAX_THREAD_NAME_LENGTH = 16;
 
 #define THREADAPI
 
-typedef void *THREADAPI_RETURN_TYPE;
-typedef std::function<THREADAPI_RETURN_TYPE THREADAPI()> ThreadFunction;
+using THREAD_RETURN_TYPE = void *;
+using ThreadFunction     = std::function<THREAD_RETURN_TYPE THREADAPI()>;
```
Comment (jpeach) [inline iocore/eventsystem/I_Thread.h]: The code changes in this commit should be make in a separate commit.

File: doc/developer-guide/architecture/threads-and-events.en.rst (hunk 7 above)
Comment (SolidWallOfCode) [inline doc/developer-guide/architecture/threads-and-events.en.rst]: I'm not really sure of the situation there. This is in `I_Net.h`
```
/*
 * Net system uses event threads
 * so, the net thread group id is the event thread group id
 */

#define ET_NET ET_CALL
```
In practice these are always referred to as "ET_NET" threads, so it looks like "ET_CALL" is the obsolete value.

File: doc/developer-guide/architecture/threads-and-events.en.rst (hunk 9 above)
Comment (SolidWallOfCode) [inline doc/developer-guide/architecture/threads-and-events.en.rst]: Heh, the next commit removes that from the documentation.

**Comment by SolidWallOfCode** [comment]
I'm unclear on how the documentation would be usefully split out, except as entirely separate. It's not really a documentation update as the current documentation has nothing at all on this subject, the content in this PR is completely new. I have kept them bundled to make the code changes easier to understand, rather than having to consult a separate PR to read much of the logical structuring of the code.

**Comment by zwoop** [comment]
[approve ci autest]

File: iocore/eventsystem/I_Thread.h (hunk 10 above)
Comment (SolidWallOfCode) [inline iocore/eventsystem/I_Thread.h]: #2075

Comment (jpeach) [inline iocore/eventsystem/I_Thread.h]: Thanks!

**Comment by SolidWallOfCode** [comment]
Waiting on two other PRs to go, once those are done I will rebase to clear conflicts.

**Comment by SolidWallOfCode** [comment]
This is now stripped down about as far as possible, time to make a judgement.

**Comment by SolidWallOfCode** [comment]
[ci approve autest]

**Comment by SolidWallOfCode** [comment]
[ci approve autest]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Comment by SolidWallOfCode** [comment]
[approve ci]
